lighthouse-plugin-publisher-ads/audits/script-injected-tags.js

```diff
@@ -98,39 +93,71 @@ class ScriptInjectedTags extends Audit {
-      {trace, devtoolsLog}, context);
+    const networkRecords = await NetworkRecords.request(devtoolsLog, context);
+    const tagReqs = networkRecords.filter(
+      (req) => STATICALLY_LOADABLE_TAGS.find((t) => req.url.match(t)));
+
+    const criticalRequests = getAdCriticalGraph(
+      networkRecords, trace.traceEvents);
+    // Find critical scripts that were loaded by inline scripts, which could
+    // have been loaded statically.
+    for (const req of criticalRequests) {
+      if (req.resourceType !== 'Script') {
+        // Don't count resources that aren't scripts.
+        continue;
+      }
+      if (req.initiator.type !== 'script') {
+        // Don't count resources that weren't loaded by inline scripts.
+        continue;
+      }
+      const initiators = PageDependencyGraph.getNetworkInitiators(req);
+      if (initiators.length !== 1 || initiators[0] !== req.documentURL) {
+        // We can't recommend static loading of requests that depend on other
+        // scripts.
+        continue;
+      }
+      tagReqs.push(req);
+    }
+
+    if (!tagReqs.length) {
+      return auditNotApplicable.NoTag;
+    }
+
+    const seenUrls = new Set();
+    const tableView = [];
+
+    const scriptTimes =
+      await getScriptEvaluationTimes(trace, devtoolsLog, context);
+    for (const tag of tagReqs) {
+      if (seenUrls.has(tag.url)) continue;
+      seenUrls.add(tag.url);
+      const relatedTags = tagReqs.filter((t) => t.url === tag.url);
 
-    const closure = getTransitiveClosure(mainDocumentNode, isGptAdRequest);
-    if (!closure.requests.length) {
-      return auditNotApplicable.NoAdRelatedReq;
+      const numStatic = array.count(
+        relatedTags, (t) => t.initiator.type === 'parser' && !t.isLinkPreload);
+      if (numStatic === 0) {
+        const loadTime = scriptTimes.get(tag.url) || 0;
+        if (loadTime < MINIMUM_LOAD_TIME_MS) {
+          continue;
+        }
+        tableView.push({
+          url: tag.url,
+          loadTime,
+        });
+      }
     }
-    const injectedBlockingRequests = closure.requests
-        .filter((r) =>
-          r.resourceType == 'Script' && r.initiator.type == 'script')
-        .filter((r) =>
-          r.initiatorRequest && r.initiatorRequest.url == r.documentURL);
-
-    /** @type {Map<NetworkRequest, NodeTiming>} */
-    const timings = await getTimingsByRecord(
-      trace, devtoolsLog, context);
-    const tableView = injectedBlockingRequests.map((req) =>
-      Object.assign({}, timings.get(req), {
-        request: req.url,
-        lineNumber: req.initiator.stack.callFrames[0].lineNumber,
-      }));
-    tableView.sort((a, b) => a.startTime - b.startTime);
+    tableView.sort((a, b) => a.loadTime - b.loadTime);
 
     const failed = tableView.length > 0;
     return {
+      displayValue: failed ?
+        str_(UIStrings.failureDisplayValue, {tags: tableView.length}) : '',
+      score: Number(!failed),
```

Comment:
Should this be a percent of the total or something?

Reply:
Agreed. Most scores should be non-binary--but I'll leave it as-is for now.

Maybe the score should be based on load time rather than number of requests? The trouble is that load time doesn't always corresponds to actual expected speed improvement.